doc/developers/contributing.rst

```diff
@@ -734,10 +736,22 @@ In general have the following in mind:
     5. Specify ``dataframe`` when "frame-like" features are being used, such
        as the column names.
     6. When specifying the data type of a list, use ``of`` as a delimiter:
-       ``list of int``.
+       ``list of int``. When the parameter supports arrays giving details about
+       the shape and/or data type and a list of such arrays, you can use one
+       of:
+
+      * ``array-like of shape (n_samples,) or list thereof``;
+      * ``array-like of shape (n_samples,) or list of these``;
+      * ``array-like of shape (n_samples,) or list of such arrays``.
+
     7. When specifying the dtype of an ndarray, use e.g. ``dtype=np.int32``
        after defining the shape:
-       ``ndarray of shape (n_samples,), dtype=np.int32``.
+       ``ndarray of shape (n_samples,), dtype=np.int32``. You can specify
+       multiple dtype as a set:
+       ``array-like of shape (n_samples,), dtype={np.float64, np.float32}``.
+       If one wants to mention arbitrary precision, use `integral` and
+       `floating` rather than the Python dtype `int` and `float`. When both
+       `int` and `floating` are supported, one can use `dtype=numeric`.
```

Comment:
I think we should just recommend to omit an explicit mention of the dtype in parameter docs whenever the method or function accepts arrays with any numeric dtype and only mention it in specific cases where a specific precision is expected or more efficient but then this is better be done in the body of the attribute/parameter doc rather than the oneliner header.

I believe this is what we currently do most of the time.